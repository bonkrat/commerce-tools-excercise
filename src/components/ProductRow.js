import React from "react";
import PropTypes from "prop-types";
import moment from "moment";
import { useHistory } from "react-router-dom";

/**
 * Renders a table row of product data.
 */
const ProductRow = ({
  index,
  id,
  name,
  productType,
  published,
  categories,
  createdAt
}) => {
  const history = useHistory();
  const coloredRowClass = index % 2 ? "bg-gray-100 " : "";
  return (
    <tr
      className={coloredRowClass + "cursor-pointer hover:bg-gray-200"}
      onClick={() => history.push("/" + id)}
    >
      <td className="border px-4 py-2">{name}</td>
      <td className="border px-4 py-2">{productType}</td>
      <td className="border px-4 py-2">
        {published ? "Published" : "Unpublished"}
      </td>
      <td className="border px-4 py-2">
        {categories.reduce((acc, curr) => {
          acc = acc.length ? (acc += ", " + curr.name) : curr.name;
          return acc;
        }, "")}
      </td>
      <td className="border px-4 py-2">
        {moment(createdAt).format("MMMM Do YYYY, h:mm:ss a")}
      </td>
    </tr>
  );
};

ProductRow.propTypes = {
  /**
   * The index of the current row. Controls the background color of the row.
   */
  index: PropTypes.number,
  /**
   * The id of the product.
   */
  id: PropTypes.string,
  /**
   * The name of the product.
   */
  name: PropTypes.string,
  /**
   * The type of product.
   */
  productType: PropTypes.string,
  /**
   * Whether the product is published.
   */
  published: PropTypes.bool,
  /**
   * The categories assigned to the product.
   */
  categories: PropTypes.arrayOf(
    PropTypes.shape({
      /**
       * The name of the category.
       */
      name: PropTypes.string
    })
  ),
  /**
   * The creation date of the product.
   */
  createdAt: PropTypes.string
};

export default ProductRow;
