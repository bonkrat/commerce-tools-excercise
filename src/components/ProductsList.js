import React from "react";
import PropTypes from "prop-types";
import ProductRow from "./ProductRow";
import { useQuery } from "@apollo/react-hooks";
import { gql } from "apollo-boost";
import Paging from "../containers/PagingContainer";
import QueryPredicate from "../containers/QueryPredicateContainer";
import LinearProgress from "@material/react-linear-progress";
import { times } from "lodash";

/**
 * Builds a search query for the commercetools GraphQL API.
 *
 * @param {string} predicate The search query predicate for searching for results with the GraphQL query.
 * @param {number} offset The paging offset (pageNumber).
 * @param {number} limit The number of results (pageSize).
 */
export const buildProductsQuery = (predicate, offset, limit) => gql`
  query ProductsQuery {
    products (offset: ${offset}, limit: ${limit}, ${
  predicate ? `where:"${predicate}"` : ""
} ){
      total
      results {
        id
        createdAt
        productType {
          name
        }
        masterData {
          published
          current {
            name(locale: "en")
            description(locale: "en")
            categories {
              name(locale: "en")
            }
          }
        }
      }
    }
  }
`;

/**
 * Placeholder rows for display while the table is loading.
 */
export const PlaceholderRows = ({ pageSize }) =>
  times(pageSize, index => (
    <tr key={index} className={`${index % 2 ? "bg-gray-100" : ""}`}>
      <td className="border px-4 py-6"></td>
      <td className="border px-4 py-6"></td>
      <td className="border px-4 py-6"></td>
      <td className="border px-4 py-6"></td>
      <td className="border px-4 py-6"></td>
    </tr>
  ));

PlaceholderRows.propTypes = {
  /**
   * The number of items displayed in the table page.
   */
  pageSize: PropTypes.number
};

/**
 * A table of product results.
 */
export const Table = ({ children }) => (
  <table className="table-fixed bg-white w-full">
    <thead className="bg-gray-100">
      <tr>
        <th className="w-1/4 px-4 py-2">Product Name</th>
        <th className="w-1/16 px-4 py-2">Product Type</th>
        <th className="w-1/16 px-4 py-2">Status</th>
        <th className="w-1/4 px-4 py-2">Categories</th>
        <th className="w-1/4 px-4 py-2">Created</th>
      </tr>
    </thead>
    <tbody>{children}</tbody>
  </table>
);

Table.propTypes = {
  children: PropTypes.oneOfType([PropTypes.array, PropTypes.node])
};

/**
 * Displays a table of product data by querying the commercetools GraphQL API.
 */
export default () => {
  const { offset, pageSize, setLastPage } = Paging.useContainer();
  const { queryPredicate } = QueryPredicate.useContainer();
  const { loading, error, data } = useQuery(
    buildProductsQuery(queryPredicate, offset, pageSize)
  );

  // Bug included if pageSize is the last page is the length of pageSize.
  data?.products?.results?.length < pageSize
    ? setLastPage(true)
    : setLastPage(false);

  return (
    <React.Fragment>
      {error && "There was an error"}
      <LinearProgress
        indeterminate
        className={loading ? "opacity-100" : "opacity-0"}
      />
      <Table>
        {loading ? (
          <PlaceholderRows pageSize={pageSize} />
        ) : (
          data?.products?.results.map((product, index) => {
            const {
              id,
              createdAt,
              productType: { name: productType },
              masterData: { published, current }
            } = product;
            return (
              <ProductRow
                index={index}
                key={id}
                id={id}
                productType={productType}
                createdAt={createdAt}
                published={published}
                {...current}
              />
            );
          })
        )}
      </Table>
    </React.Fragment>
  );
};
