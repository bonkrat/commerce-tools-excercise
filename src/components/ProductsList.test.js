import React from "react";
import ProductsList, {
  PlaceholderRows,
  Table,
  buildProductsQuery
} from "./ProductsList";
import { shallow } from "enzyme";
import { useQuery } from "@apollo/react-hooks";
import Paging from "../containers/PagingContainer";
import QueryPredicate from "../containers/QueryPredicateContainer";
import LinearProgress from "@material/react-linear-progress";
import ProductRow from "./ProductRow";
import { times } from "lodash";

jest.mock("../containers/PagingContainer", () => ({
  useContainer: jest.fn(() => ({
    offset: 0,
    pageSize: 10,
    setLastPage: jest.fn()
  }))
}));

jest.mock("../containers/QueryPredicateContainer", () => ({
  useContainer: jest.fn(() => ({
    queryPredicate: "foopredicate"
  }))
}));

jest.mock("@apollo/react-hooks", () => ({
  useQuery: jest.fn(() => ({
    loading: false,
    error: false,
    data: {
      products: {
        results: [
          {
            id: "foo",
            createdAt: "12345",
            productType: {
              name: "bar"
            },
            masterData: {
              published: true,
              current: {}
            }
          }
        ]
      }
    }
  }))
}));

describe("buildProductsQuery", () => {
  it("builds a query for searching products", () => {
    expect(buildProductsQuery("foobar", 5, 5)).toMatchSnapshot();
  });
});

describe("PlaceholderRow", () => {
  it("prints a page of placeholder rows", () => {
    expect(shallow(<PlaceholderRows pageSize={5} />).find("td").length).toBe(
      25
    );
  });
  it("sets the background color on the rows", () => {
    const wrapper = shallow(<PlaceholderRows pageSize={5} />).find("tr");
    times(5, index => {
      if (index % 2) {
        expect(wrapper.at(index).prop("className")).toBe("bg-gray-100");
      } else {
        expect(wrapper.at(index).prop("className")).toBe("");
      }
    });
  });
});

describe("Table", () => {
  it("renders children", () => {
    expect(
      shallow(<Table>{["foobar"]}</Table>)
        .findWhere(n => n.text() === "foobar")
        .exists()
    ).toBeTruthy();
  });
});

describe("ProductsList", () => {
  afterEach(() => {
    useQuery.mockClear();
    Paging.useContainer.mockClear();
    QueryPredicate.useContainer.mockClear();
  });

  it("prints rows of data", () => {
    const wrapper = shallow(<ProductsList />);
    expect(wrapper.find(ProductRow).props()).toMatchObject({
      index: 0,
      createdAt: "12345",
      id: "foo",
      productType: "bar",
      published: true
    });
  });

  it("builds a product query", () => {
    shallow(<ProductsList />);
    expect(useQuery).toHaveBeenCalledWith(
      buildProductsQuery("foopredicate", 0, 10)
    );
  });

  it("sets the last page if there is no more paging data", () => {
    Paging.useContainer.mockImplementation(() => ({
      offset: 0,
      pageSize: 1,
      setLastPage: jest.fn()
    }));

    shallow(<ProductsList />);

    expect(
      Paging.useContainer.mock.results[0].value.setLastPage
    ).toHaveBeenCalledWith(false);

    Paging.useContainer.mockReset();

    Paging.useContainer.mockImplementation(() => ({
      offset: 0,
      pageSize: 25,
      setLastPage: jest.fn()
    }));

    shallow(<ProductsList />);

    expect(
      Paging.useContainer.mock.results[0].value.setLastPage
    ).toHaveBeenCalledWith(true);
  });

  it("shows a message if there is an error", () => {
    useQuery.mockImplementation(() => ({
      loading: false,
      error: true,
      data: { products: { results: [] } }
    }));

    const wrapper = shallow(<ProductsList />);
    expect(
      wrapper.findWhere(n => n.text() === "There was an error").exists()
    ).toBeTruthy();
  });

  it("shows a loading indicator when fetching data", () => {
    useQuery.mockImplementation(() => ({
      loading: true,
      error: false,
      data: {
        products: {
          results: [
            {
              id: "foo",
              createdAt: "12345",
              productType: {
                name: "bar"
              },
              masterData: {
                published: true,
                current: {}
              }
            }
          ]
        }
      }
    }));

    const wrapper = shallow(<ProductsList />);
    expect(wrapper.find(LinearProgress).prop("className")).toBe("opacity-100");
    expect(wrapper.find(LinearProgress).prop("indeterminate")).toBeTruthy();
    expect(wrapper.find(PlaceholderRows).exists()).toBeTruthy();
    expect(wrapper.find(PlaceholderRows).prop("pageSize")).toBe(25);
  });
});
