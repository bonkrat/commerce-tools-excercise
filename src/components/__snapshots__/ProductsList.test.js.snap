// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildProductsQuery builds a query for searching products 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "kind": "OperationDefinition",
      "name": Object {
        "kind": "Name",
        "value": "ProductsQuery",
      },
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": undefined,
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": Object {
                  "kind": "Name",
                  "value": "offset",
                },
                "value": Object {
                  "kind": "IntValue",
                  "value": "5",
                },
              },
              Object {
                "kind": "Argument",
                "name": Object {
                  "kind": "Name",
                  "value": "limit",
                },
                "value": Object {
                  "kind": "IntValue",
                  "value": "5",
                },
              },
              Object {
                "kind": "Argument",
                "name": Object {
                  "kind": "Name",
                  "value": "where",
                },
                "value": Object {
                  "block": false,
                  "kind": "StringValue",
                  "value": "foobar",
                },
              },
            ],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "products",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "total",
                  },
                  "selectionSet": undefined,
                },
                Object {
                  "alias": undefined,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "results",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "createdAt",
                        },
                        "selectionSet": undefined,
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "productType",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "name",
                              },
                              "selectionSet": undefined,
                            },
                          ],
                        },
                      },
                      Object {
                        "alias": undefined,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "masterData",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "published",
                              },
                              "selectionSet": undefined,
                            },
                            Object {
                              "alias": undefined,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "current",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [
                                      Object {
                                        "kind": "Argument",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "locale",
                                        },
                                        "value": Object {
                                          "block": false,
                                          "kind": "StringValue",
                                          "value": "en",
                                        },
                                      },
                                    ],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "name",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [
                                      Object {
                                        "kind": "Argument",
                                        "name": Object {
                                          "kind": "Name",
                                          "value": "locale",
                                        },
                                        "value": Object {
                                          "block": false,
                                          "kind": "StringValue",
                                          "value": "en",
                                        },
                                      },
                                    ],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "description",
                                    },
                                    "selectionSet": undefined,
                                  },
                                  Object {
                                    "alias": undefined,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "categories",
                                    },
                                    "selectionSet": Object {
                                      "kind": "SelectionSet",
                                      "selections": Array [
                                        Object {
                                          "alias": undefined,
                                          "arguments": Array [
                                            Object {
                                              "kind": "Argument",
                                              "name": Object {
                                                "kind": "Name",
                                                "value": "locale",
                                              },
                                              "value": Object {
                                                "block": false,
                                                "kind": "StringValue",
                                                "value": "en",
                                              },
                                            },
                                          ],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "name",
                                          },
                                          "selectionSet": undefined,
                                        },
                                      ],
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                        },
                      },
                    ],
                  },
                },
              ],
            },
          },
        ],
      },
      "variableDefinitions": Array [],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 424,
    "start": 0,
  },
}
`;
